[9.0][MIG] l10n_es_aeat_sii. pasar modulo version 8 a 9

RamonGuiuGou opened the pull request:
Traspaso del modulo l10n_es_aeat_sii de la versión 8 hacia la 9.

pedrobaeza commented:
@RamonGuiuGou, lo suyo sería respetar el historial de commits (aunque primero habría que racionalizarlo un poco en v8, juntando commits que no sirvan de mucho). ¿Esto incorpora además los últimos cambios hechos en v8?

RamonGuiuGou commented:
@pedrobaeza si incorpora todos los commits hechos hasta ayer por la mañana de la branca de acysos de la versión 8, los he ido migrando. Para si alguien está interesado en la versión 9 que tenga una versión más actual.
Ahora al mediodía he visto que se han aceptado varios PR's de la versión 8, cuando tengo un ratillo también los migraré.

pedrobaeza commented:
Pero ya te digo que no estás respetando la autoría de ese código. Vamos a centrarnos en esta semana en fusionar una versión válida para v8, y a partir de ahí, es transferir los commits (se puede utilizar esta guía https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-9.0), y montar encima los cambios que hagan falta ya con tu autoría.

RamonGuiuGou commented:
@pedrobaeza Vale, si lo haré siguiendo está guía para respectar la auditoria, que ahora no la respeto. Perdona mi desconocimiento.

Roodin commented:
@RamonGuiuGou Para poder seguir con el trabajo en V10 de forma correcta necesitaríamos que estuviese corregido  el historial de commits para respetar la autoria. Entiendo que debería hacerse una limpieza antes en la 8 pero no podemos esperar más para ternerlo en pruebas. ¿Alguna estimación de cuado podráis tener esto o lo preparamos nosotros?

RamonGuiuGou commented:
@Roodin El módulo tal y como esta ahora es funcional.Ya que, casi todos los PR's de la 8 están migrados. Unicamente, falta traspasarlo respetando la auditoria. Entre hoy y mañana creemos que lo podremos hacer haciendo un rebase y limpiarlo un poco.

pedrobaeza commented:
@RamonGuiuGou espera al viernes que voy a hacer limpieza de la rama 8.0 (ayer ya hice un poco eliminando merges), y si quieres vemos juntos cómo hacerlo con esta rama.

RamonGuiuGou commented:
@pedrobaeza Ok perfecto! Si mejor, vale más que la rama de la 8 este limpia. Y si té va bien nos coordinamos para hacerlo en esta rama.

pedrobaeza commented:
Pasar los nuevos cambios de la 8 a la 9 está siendo un infierno porque no habéis hecho algo tan sencillo como poner un commit inicial que fuera "esto es lo que hay de versión 8" y vuestros cambios vinieran después en sucesivos commits. La próxima vez haced el favor de hacerlo de esta forma.

RamonGuiuGou committed:
[MIG] l10n_es_aeat_sii: Migration to 9.0

pedrobaeza commented:
Ya está el PR arreglado respetando la historia y añadiendo solo los cambios mínimos para la versión. Ahora voy a pasar a refactorizar la versión 8, que trasladaré aquí, y después arreglar también algunas cosas en la migración a v9 en si (estilo, PEP8, etc).

pedrobaeza commented:
@RamonGuiuGou, no veo claro lo del redondeo en todas y cada una de las cifras, ya que así pueden salir cifras distintas a lo que hace Odoo (ya sabes, la suma de los redondeos no es lo mismo que el redondeo de la suma). De todas formas, lo que sea habría que hacerlo igual en v8.

Voy a estudiarlo detenidamente para comparar y lo que sea lo traigo a las dos versiones, ¿OK?

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_sii: Make version 9 to work

pedrobaeza commented:
Ya he hecho funcionar la versión 9.0, incluidos tests. ¿Lo tenéis en alguna BD en marcha?

RamonGuiuGou commented:
@pedrobaeza si el tema del redondeo se tendría que dar una vuelta más ya que depende de como el usuario tenga configurado el calculo de iva global o por linea... puede dar desajustes importantes en las facturas. Por eso, fue una mala forma de hacerlo. También vamos a pensar sobre el tema haber que se nos ocurre.

pedrobaeza commented:
En v8 ya se añadió un parche para ello, aunque es algo parcial: https://github.com/OCA/l10n-spain/pull/464/files#diff-1e50d7d6033ffb0d15c5823d421fc2c2R266

He visto también que la lógica de la clave de exención de los productos estaba aquí, pero no en la v8. ¿Podrías proponer la misma en v8?

RamonGuiuGou commented:
Ok, la lógica de los productos exentos la subiremos, para la v.8 y después si os parece bien la migramos.

pedrobaeza commented:
Bueno, ahora mismo ya que está bien separados los commits, estoy sincronizando ambas ramas, por lo que poniéndolo en la 8, ya lo traigo automáticamente aquí.

MouTio commented:
No estoy muy seguro de si tengo que reportar esto aquí o dónde (en issues supongo que van fallos de módulos terminados).

Al instalar el PR, el enlace a la Configuración general (Configuración -> General Settings) deja de funcionar (ahora se queda como una etiqueta sin hipervínculo).

pedrobaeza commented:
Sí, sería aquí, ya que aquí estamos "cocinándolo". Efectivamente, me había dado yo cuenta también de eso, pero no lo había cambiado aún. Lo cambiaré luego, aunque estate atento por si se me va la olla y se me olvida.

MouTio commented:
Al enviar una factura al entorno de pruebas, perfecto. Si la intento enviar al entorno de producción, me aparece el siguiente error:

Retorno SII
Codigo[4100].Error en la cabecera. El contenido del campo IDVersionSii no es válido.

He visto en los parámetros del sistema que el IDVersionSii es 0.7, pero en la AEAT dicen que en el entorno de producción sólo admiten la 1.0
Por lo tanto, es correcto que me aparezca este error en producción.

Cuando se suba la versión final, ¿El módulo cambiará automáticamente la versión de 0.7 a 1.0 el día 1 de julio o tenemos que modificar nosotros el l10n_es_aeat_sii.version de system parameters?

Y otra cosa: ¿El día 1 de julio habrá que enviar todas las facturas emitidas desde el 1 de enero pulsando el botón de "Enviar SII" una por una?

pedrobaeza commented:
La AEAT aún no ha habilitado el entorno de producción y todavía no se sabe cómo lo van a gestionar. Ya se cambiará lo que haya que cambiar.

Para lo otro, aún no se ha implementado la opción de enviar las del semestre anterior (van con una clave especial).

MouTio commented:
De acuerdo @pedrobaeza. Muchas gracias.

Un pequeño apunte:
El enlace que hay en la descripción del módulo, apunta a una URL que ya no existe.
La nueva URL es:
http://www.agenciatributaria.es/static_files/AEAT/Contenidos_Comunes/La_Agencia_Tributaria/Modelos_y_formularios/Suministro_inmediato_informacion/FicherosSuministros/Version_Produc/SII_Descripcion_ServicioWeb_es_es.pdf

MouTio commented:
Están estas dos versiones:
**Opción 1**: https://github.com/OCA/l10n-spain/tree/9.0/l10n_es_aeat_sii
**Opción 2** (PR): https://github.com/OCA/l10n-spain/pull/514

Tengo una duda bastante básica:
¿Cuál es la última versión (que funcione) para descargar? ¿La opción 1 o la opción 2?

Con la opción 2 (PR de 514), me pide instalar el módulo **account_invoice_currency**, pero el otro día no me lo pedía. ¿Esto es correcto?

pedrobaeza commented:
No hay dos opciones, sólo hay ésta. Lo otro son los commits de la 8.0 llevados a la rama principal, pero no operativos para v9.

MouTio commented:
Gracias por la aclaración, @pedrobaeza 
Entonces, sólo está la opción del PR 514.

Después de instalar las 4 dependencias:
**connector
l10n_es_aeat
account_invoice_currency
account_invoice_refund_link**

No me deja instalar el módulo del SII, da el siguiente error:
`El campo sii_header_supplier no existe`

¿Tengo que crear este campo a mano en la base de datos o estoy haciendo algo mal?

pedrobaeza commented:
Es cosa tuya de no actualizar el módulo en Odoo.

MouTio commented:
Ya he conseguido actualizar el módulo.

Ahora, cuando intento enviar alguna factura, me aparece el campo de "Descripción SII" vacío, cuando antes me ponía una barra (/) de manera automática. Aunque lo configure como automático o fijo, me sigue saliendo vacío. ¿Tengo que rellenarlos uno por uno a mano?
Además, cuando lo relleno a mano y le doy a enviar, me devuelve el error de "Factura duplicada"

pedrobaeza commented:
Pon en Cabecera SII ese '/' si no quieres otras cosas

MouTio commented:
He puesto '/' y me aparece en las nuevas facturas que creo a partir de ahora, pero en las que ya tenía creadas me las deja con el campo vacío. Puedo solucionarlo con un update masivo a través de base de datos, pero quiero asegurarme primero de que no hay otra manera.

pedrobaeza commented:
Sí, haz update masivo

MouTio commented:
He probado a desmarcar el checkbox de **entorno de test** y ahora al intentar enviar una factura al SII me aparece el siguiente error:
File "/mnt/extra-addons/l10n-spain/l10n_es_aeat_sii/models/account_invoice.py", line 712, in <lambda>
    i.period_id.date_start >= SII_START_DATE
AttributeError: 'account.invoice' object has no attribute 'period_id'

Parece que intenta acceder al campo **period_id** y éste no existe.

pedrobaeza commented:
Cierto, eso no lo hemos cambiado, pero aún no se puede activar el entorno de producción.

pedrobaeza commented:
Ya está arreglado eso.

MouTio commented:
Acabo de probar de nuevo con la penúltima versión (antes de la corrección de @pedrobaeza ) y estos son los fallos que me salen:

### Fallo 1: Error de certificados

A veces, de repente deja de funcionar lo de Enviar SII, dando el siguiente error:
`[('system library', 'fopen', 'No such file or directory'), ('BIO routines', 'FILE_CTRL', 'system lib'), ('SSL routines', 'SSL_CTX_use_certificate_file', 'system lib')]`
No sé por qué ocurre este fallo de repente, pero lo consigo solucionar yendo a Configuración -> General Settings -> SII -> Certificados SII y volviendo a introducir el mismo certificado de nuevo. Al cabo de un tiempo (unas horas, o unos días) vuelve a fallar de nuevo y tengo que introducir el certificado de nuevo.


### Fallo 2: XML incorrecto: DesgloseTipoOperacion (Facturas de Cliente)

`Codigo[4102].El XML no cumple el esquema. Falta informar campo obligatorio.: DesgloseTipoOperacion`
Lo curioso aquí es que tengo 6 facturas exactamente iguales (mismo cliente, mismo concepto y mismo importe), cada una de un mes diferente (desde enero hasta junio de 2017). Las 5 primeras se han subido sin problemas y la de junio me devuelve este error.
Si voy a la web de la AEAT y la introduzco a mano, no da ningún fallo, pero claro, en Odoo se queda como SII Failed.

### Fallo 3: DescripcionOperacion

Me aparece este error al **Enviar al SII** en alguna factura:
`1100 | Valor o tipo incorrecto del campo: DescripcionOperacion`

Si le vuelvo a dar a **Enviar al SII**, me aparece: 
`Factura duplicada`

Me gustaría saber también cómo proceder en el caso de facturas duplicadas.
¿Es correcto editar a mano para desmarcar el checkbox de **SII send failed** y poner **SII send state** = **Enviada**?

MouTio commented:
A la espera de que alguien conteste mi último mensaje, lanzo esta otra consulta:

Cuando intento crear una factura rectificativa, al darle a **Crear factura rectificativa**, me aparece el siguiente mensaje de error:
```
Odoo Aviso - Error de validación

The operation cannot be completed, probably due to the following:
- deletion: you may be trying to delete a record while other records still reference it
- creation/update: a mandatory field is not correctly set

[object with reference: sii_registration_key - sii.registration.key]
```

Entiendo que hace falta rellenar el campo **sii_registration_key**, pero no me aparece por ningún sitio para rellenar. Sólo me aparece el campo **SII Refund Type**

pedrobaeza commented:
@MouTio eso es que no tienes establecido un valor para ese campo en la posición fiscal correspondiente.

MouTio commented:
@pedrobaeza En la factura original (la que rectificamos), el campo "Posición Fiscal" (en la pestaña de Otra Información) está establecido a **Régimen Nacional**.

En la rectificativa no me aparece nada de posición fiscal.

pedrobaeza commented:
No es lo que he dicho en mi comentario anterior. Campo "Clavo de registro SII" en la posición fiscal.

MouTio commented:
Vale, es dentro de la posición fiscal, lo había entendido mal. Ya no da el error. Muchas gracias

MouTio commented:
@pedrobaeza 

En la línea 202 del fichero models/account_invoice.py, en lugar de esto:
`    self.onchange_fiscal_position_l10n_es_aeat_sii()
`
debería ir esto:
`    self.onchange_fiscal_position_id_l10n_es_aeat_sii()
`

ya que la primera función no existe.

Saludos

pedrobaeza commented:
Cierto, lo cambio.

pedrobaeza commented:
Hecho

pedrobaeza committed:
[IMP] l10n_es_aeat_sii: Change tax computation to invoice level

Odoo v9 already knows which tax generates each invoice tax line, so
we get with this:

* Avoid to make roundings that are not accurate.
* If there are manual changes in supplier invoices taxes, now we
  will be able to notify them and prevent not matching errors.
* Less difficult algorithms for getting tax amount totals.

pedrobaeza commented:
@Roodin @RamonGuiuGou @MouTio @omar7r acabo de introducir un cambio grande en este PR para aprovechar las características de la nueva versión de Odoo, y es para no obtener los impuestos calculándolos de las líneas, si no utilizando el resumen a pie de factura, que ya incorpora información del impuesto originario. De esta forma, tenemos:

- Se quita de un plumazo toda la lógica de redondeos extraños.
- Se simplifica el código, al no tener que obtener agregados.
- En las facturas de proveedor, si se cambiaba manualmente el importe de la cuota para casar con lo que mandaba el proveedor, en v8 se pierde ese ajuste y supondrá un problema. Ahora aquí ya no existe ese problema.

pedrobaeza commented:
Estoy ahora haciendo pruebas reales para ver su funcionamiento correcto, así que puede ser que haya algún cambio.

pedrobaeza commented:
Bueno, pues finalmente resulta que sí que había que hacer un redondeo sobre los datos de la tabla account.invoice.tax, porque los guarda sin redondear y luego los presenta en pantalla redondeados, pero lo bueno es que ahora no hay que plantearse si redondeo de la suma o suma de redondeos: se redondea una vez y ya está.

De paso, he mejorado la visualización de los resultados del SII para tenerlo en 2 pestañas, la cual una es técnica, y además se guarda el contenido de lo que se ha enviado al SII para cualquier depuración que se quiera hacer.

Dejo mañana para que echéis un vistazo por si encontráis cualquier cosa, y si no hay nada en contra, el martes mergeo este PR para ya tener una versión estable en la rama principal, y después vamos con la 10.0!

MouTio commented:
Algo ocurre con la nueva versión. 
Intento modificar una factura para darle valor al campo **Fecha factura**, y al guardar me da el siguiente error:

`No puede cambiar la fecha de factura de una factura enviada al SII. Debe cancelar la factura y crear una nueva con la fecha correcta`

Por supuesto, esta factura no ha sido enviada al SII todavía.

**EDITO**: Si pongo la versión del PR del SII que tenía el viernes, sí que me deja editar el campo de **Fecha factura** sin darme ningún error. Por tanto, queda confirmado que es problema del último PR.

MouTio commented:
@pedrobaeza creo que sé dónde está el fallo.

En la línea 156 de models/account_invoice.py tenemos esto:

```
for invoice in self.filtered(lambda x: x.state != 'not_sent'):
    if 'date_invoice' in vals:
        raise exceptions.Warning(
            _("You cannot change the invoice date of an invoice "
              "already registered at the SII. You must cancel the "
              "invoice and create a new one with the correct date")
        )
```

Sin embargo, creo que en lugar de **state**, lo correcto sería comprobar el estado del campo **sii_state**:

`for invoice in self.filtered(lambda x: x.sii_state != 'not_sent'):`

pedrobaeza commented:
Sí, está corregido desde hace una hora y pico :wink:  Fallo mío que se me coló un typo. Sigo afinando permisos de todas formas. Estate atento a los últimos cambios.

MouTio commented:
@pedrobaeza 
¿No se puede hacer nada con el tema de los certificados?

Cada varios días deja de funcionar y me toca ir a Configuración -> SII -> Certificados SII -> y darle a obtener claves, introducir de nuevo la contraseña y poner un nombre a la carpeta.

¿Voy a tener que estar haciendo esto continuamente?.
Alguna solución alternativa debe haber para que no esté cada 2x3 con el error de certificado.

pedrobaeza commented:
@MouTio eso es cosa tuya de tu instalación y cómo la tengas hecha. Yo no tengo ese problema.

MouTio commented:
@pedrobaeza Pensaba que era un fallo de algo de la programación del módulo. Sabiendo que no es así y que a ti no te pasa, ya investigo yo qué puede ser. 

Utilizo docker y seguramente tenga que incluir el certificado en el docker antes de subirlo.
Gracias.

pedrobaeza commented:
No digas más: tienes que tener el datadir en un volumen persistente igual que tienes la BD o el filestore. Echa un vistazo a nuestra solución Docker: https://github.com/Tecnativa/docker-odoo-base

MouTio commented:
Efectivamente, el problema estaba en que debía estar en un directorio persistente como el filestore.

Haciendo pruebas, he encontrado otros fallos menores, en la pantalla de certificados SII. Los pongo a continuación junto con los pasos para reproducirlos:

**Fallo 1: Fecha final se rellena con el correo del usuario actual.**
Rellenar el campo de _nombre_ y darle a _Seleccionar_ el certificado. Al darle a _Obtener Claves_, se rellena el campo de _Fecha final_ con el correo del usuario actual.

**Fallo 2: Al darle al botón de limpiar, el formulario queda inutilizable.**
Introducimos un certificado, y le damos al botón de _Limpiar_. El fichero se elimina, pero ya no podemos meter ningún otro. Sí que deja seleccionar otro fichero pero no aparece seleccionado y de hecho no está seleccionado porque si pulsamos en Guardar, nos dice que el campo Archivo es inválido.

**Fallo 3: Excepción al pulsar en Guardar como.**
Si seleccionamos un certificado y antes de darle a _Obtener Claves_ pulsamos en _Guardar como_, aparece este error:
```
Odoo Error del cliente
Uncaught SyntaxError: Unexpected token T in JSON at position 0
https://my_company.com/web#view_type=form&model=l10n.es.aeat.sii&menu_id=279&action=351:1
Traceback:
SyntaxError: Unexpected token T in JSON at position 0
    at JSON.parse (<anonymous>)
    at HTMLIFrameElement.<anonymous> (https://my_company.com/web/content/437-17d551a/web.assets_common.js:2554:379)
    at HTMLIFrameElement.dispatch (https://my_company.com/web/content/437-17d551a/web.assets_common.js:740:451)
    at HTMLIFrameElement.elemData.handle (https://my_company.com/web/content/437-17d551a/web.assets_common.js:713:173)
```


A parte de estos fallos, tengo 2 dudas: 

**Cuestión 1**: ¿Sirven para algo los campos de Fecha de Inicio y Fecha final?
**Cuestión 2**: El fichero generado al darle a Guardar como, ¿para qué sirve exactamente?

pedrobaeza commented:
Los errores que dices son todos relacionados con el widget binary de Odoo y no con el desarrollo en si, pero vamos, a mí no me ocurren, por lo que es de nuevo de tu instalación o del navegador utilizado.

Esos campos al final no sirven para nada, la verdad, pero bueno, era por no trabajar en otro frente.

Por cierto, estaría bien que te identificaras con nombre y empresa en lugar de un nickname extraño y sin foto.

pedrobaeza commented:
Bueno, tal como dije, paso a mergear el módulo para tenerlo ya en la rama principal en una versión lo suficientemente estable, que aunque tenemos un par de incidencias abiertas, cuando las resolvamos lo traemos a esta versión por cherry-pick.
